File: haskell/src/Projectile.hs
module Projectile where

import Tuples

{-
# projectile starts one unit above the origin.
# velocity is normalized to 1 unit/tick.
p ← projectile(point(0, 1, 0), normalize(vector(1, 1, 0)))

# gravity -0.1 unit/tick, and wind is -0.01 unit/tick.
e ← environment(vector(0, -0.1, 0), vector(-0.01, 0, 0))

function tick(env, proj)
  position ← proj.position + proj.velocity
  velocity ← proj.velocity + env.gravity + env.wind
  return projectile(position, velocity)
end function

-}

data Projectile = Projectile { position :: Tuple
                             , velocity :: Tuple }
                  deriving(Show)

data Environment = Environment { gravity :: Tuple
                               , wind    :: Tuple }
                   deriving(Show)

newtype Gravity = Gravity Double
                deriving (Show)

newtype Wind = Wind Double

tick :: Environment -> Projectile -> Projectile
tick env proj = let pos = (position proj) `add` (velocity proj)
                    vel = (velocity proj) `add` (gravity env) `add` (wind env)
                in Projectile pos vel

launch' :: Environment -> Projectile -> [Projectile] -> [Projectile]
launch' env p@(Projectile (Tuple _ y _ _) _) xs
  | y < 0     = xs
  | otherwise = p : launch' env (tick env p) xs

launch :: Gravity -> Wind -> [Projectile]
launch (Gravity g) (Wind w) = let env  = Environment (vector 0 g 0) (vector w 0 0) 
                                  proj = Projectile (point 0 1 0) (norm (vector 1 1 0))
                              in launch' env proj []
                                                  
launchresult = let g = (Gravity (-0.5))
                   w = (Wind (-0.01))
               in launch g w
                  

File: rust/src/bin/projectile.rs
use ray_tracer::canvas::Canvas;
use ray_tracer::color::Color;
use ray_tracer::vector::{Point, Vector};
use std::fs::File;
use std::io::Write;
use std::path::Path;

#[derive(Debug)]
struct Projectile {
    position: Point,
    velocity: Vector,
}

struct Environment {
    gravity: Vector,
    wind: Vector,
}

fn tick(env: &Environment, proj: &Projectile) -> Projectile {
    Projectile {
        position: &proj.position + &proj.velocity,
        velocity: &proj.velocity + &env.gravity + &env.wind,
    }
}

fn launch(g: f32, w: f32) -> Vec<Projectile> {
    let env = Environment {
        gravity: Vector::new(0., g, 0.),
        wind: Vector::new(w, 0., 0.),
    };
    let proj = Projectile {
        position: Point::new(0., 1., 0.),
        velocity: (Vector::new(1., 1.8, 0.)).norm() * 11.25,
    };

    let mut ps = vec![proj];

    while ps.last().unwrap().position.y > 0. {
        let next = tick(&env, ps.last().unwrap());
        if next.position.y > 0. {
            ps.push(next)
        } else {
            return ps;
        }
    }

    ps
}

fn projectile_into_canvas(ps: &Vec<Projectile>) -> Canvas {
    let mut canvas = Canvas::new(900, 550);
    let color = Color::new(1., 0.8, 0.6);

    for Projectile { position, .. } in ps {
        println!(
            "Wrote : {:?}-{:?}",
            position.x.floor() as usize,
            (550. - position.y).floor() as usize
        );
        canvas.write_pixel(
            position.x.floor() as usize,
            (549. - position.y).floor() as usize,
            color.clone(),
        )
    }

    canvas
}

fn main() {
    let projectile_path = launch(-0.1, -0.01);

    println!("{:?}", projectile_path);

    let ppm = projectile_into_canvas(&projectile_path).to_ppm();

    let path = Path::new("projectile.ppm");

    let mut file = match File::create(&path) {
        Err(why) => panic!("could not create file {}: {} ", path.display(), why),
        Ok(file) => file,
    };

    if let Err(why) = file.write_all(ppm.as_bytes()) {
        panic!("could not write file {}: {} ", path.display(), why)
    }

    println!("Projectile done!");
}
